src/envoy/auth/config.cc

```diff
@@ -0,0 +1,162 @@
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "config.h"
+
+#include "common/filesystem/filesystem_impl.h"
+#include "common/json/json_loader.h"
+#include "envoy/json/json_object.h"
+#include "envoy/upstream/cluster_manager.h"
+
+#include "rapidjson/document.h"
+
+#include <chrono>
+#include <fstream>
+#include <iostream>
+#include <memory>
+#include <string>
+#include <vector>
+
+namespace Envoy {
+namespace Http {
+namespace Auth {
+
+void AsyncClientCallbacks::onSuccess(MessagePtr &&response) {
+  std::string status = response->headers().Status()->value().c_str();
+  if (status == "200") {
+    ENVOY_LOG(debug, "AsyncClientCallbacks [cluster = {}]: success",
+              cluster_->name());
+    std::string body;
+    if (response->body()) {
+      auto len = response->body()->length();
+      body = std::string(static_cast<char *>(response->body()->linearize(len)),
+                         len);
+    } else {
+      ENVOY_LOG(debug, "AsyncClientCallbacks [cluster = {}]: body is null",
+                cluster_->name());
+    }
+    cb_(true, body);
+  } else {
+    ENVOY_LOG(debug,
+              "AsyncClientCallbacks [cluster = {}]: response status code {}",
+              cluster_->name(), status);
+    cb_(false, "");
+  }
+}
+void AsyncClientCallbacks::onFailure(AsyncClient::FailureReason) {
+  ENVOY_LOG(debug, "AsyncClientCallbacks [cluster = {}]: failed",
+            cluster_->name());
+  cb_(false, "");
+}
+
+void AsyncClientCallbacks::Call(const std::string &uri) {
+  ENVOY_LOG(debug, "AsyncClientCallbacks [cluster = {}]: {} {}",
+            cluster_->name(), __func__, uri);
+  // Example:
+  // uri  = "https://example.com/certs"
+  // pos  :          ^
+  // pos1 :                     ^
+  // host = "example.com"
+  // path = "/certs"
+  auto pos = uri.find("://");
+  pos = pos == std::string::npos ? 0 : pos + 3;  // Start position of host
+  auto pos1 = uri.find("/", pos);
+  if (pos1 == std::string::npos) pos1 = uri.length();
+  std::string host = uri.substr(pos, pos1 - pos);
+  std::string path = "/" + uri.substr(pos1 + 1);
+
+  MessagePtr message(new RequestMessageImpl());
+  message->headers().insertMethod().value().setReference(
+      Http::Headers::get().MethodValues.Get);
+  message->headers().insertPath().value(path);
+  message->headers().insertHost().value(host);
+
+  cm_.httpAsyncClientForCluster(cluster_->name())
+      .send(std::move(message), *this, timeout_);
+}
+
+IssuerInfo::IssuerInfo(Json::Object *json) {
+  ENVOY_LOG(debug, "IssuerInfo: {}", __func__);
+  if (json->hasObject("name") && json->hasObject("pubkey")) {
+    name_ = json->getString("name");
```

Comment:
`getString(string)` may throw exception when the object was not a string, so use `getString(string, string)` instead.

Reply:
fixed